design/external-references.md

```diff
@@ -0,0 +1,394 @@
+
+### Deserialization
+
+To support deserialization, classes must also include a public static `deserializeXxx` method which reads the object from a deserialization context and returns an object that implements the resource type interface:
+
+> The reason we indicate the type in the method name is because static methods in JavaScript are inherited, so we can differentiate between `ExtraBucket.deserializeExtraBucket` and `ExtraBucket.deserizlizeBucket`.
+
+```ts
+class MyResource extends Construct implements ISerializable {
+  static deserializeMyResource(ctx: IDeserializationContext): IMyResource;
+}
+```
+
+The deserialization context is an object that implements the following interface:
+
+```ts
+interface IDeserializationContext {
+  scope: Construct;
+  id: string;
+  readString(key: string, options?: ReadStringOptions): string;
+  readObject(key: string): IDeserializationContext;
+}
+
+interface ReadStringOptions {
+  allowUnresolved?: boolean; /** @default true */
+}
+```
+
+`readString(key)` can be used to read values stored by `writeString`. and `readObject(key)` returns a deserialization context for composite deserialization written via `writeObject`.
+
+The `allowUnresolved` option for `readString` can be used by constructs to indicate that returned value must be a resolved value (i.e. not a token). This implies, for example, that when importing this value, users cannot use the `resolveType: Deployment`  option (REQ8).
+
+Since `deserializeXxx` will need to create new construct objects, the deserialization context will supply a consistent `scope` and `id` which can be used to instantiate a construct object that represents this object. For example, `scope` can be mapped to the current `Stack` and `id` can be mapped to `exportName` which is ensured to be unique within the environment (and therefore, the current stack).
+
+Implementers of `deserializeXxx` should check if a construct with `id` already exists within `scope` and return it instead of instantiating and new object (REQ6).
+
+---
+
+Here's an example that demonstrates how serialization and deserialization can be implemented for an application load balancer (ALB), which is a composite resource. As you can see, ALBs encapsulate a security group, which is serialized and deserialized together with the ALB itself.
+
+```ts
+class ApplicationLoadBalancer {
+  serialize(ctx: ISerializationContext): void {
+    ctx.writeString('LoadBalancerArn', this.loadBalancerArn);
+    ctx.writeObject('SecurityGroup', this.securityGroup);
+  }
+
+  static deserializeApplicationLoadBalancer(ctx: IDeserializationContext): IApplicationLoadBalancer {
+    const exists = ctx.scope.findChild(ctx.id);
```

Comment:
Feels like `ctx` has all the information that is needed to get the existing entry if it exists - why isn't this a specific method on the context, instead of forcing everyone to do it themselves? Also, the code sample lacks the forced up-casting.